plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1236,39 +1236,59 @@ public void executeRemoveOrphanFiles(ConnectorSession session, IcebergTableExecu
-    {
-        ImmutableSet<String> manifests = stream(table.snapshots())
-                .flatMap(snapshot -> snapshot.allManifests(table.io()).stream())
-                .map(ManifestFile::path)
-                .collect(toImmutableSet());
-        List<String> manifestLists = ReachableFileUtil.manifestListLocations(table);
-        List<String> otherMetadataFiles = concat(
-                metadataFileLocations(table, false).stream(),
-                Stream.of(versionHintLocation(table)))
-                .collect(toImmutableList());
-        Set<String> validMetadataFiles = concat(manifests.stream(), manifestLists.stream(), otherMetadataFiles.stream())
-                .map(IcebergMetadata::fileName)
-                .collect(toImmutableSet());
-        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, validMetadataFiles, "metadata");
+        for (Snapshot snapshot : table.snapshots()) {
+            if (snapshot.manifestListLocation() != null) {
+                validMetadataFilePaths.add(snapshot.manifestListLocation());
+            }
+
+            for (ManifestFile manifest : snapshot.allManifests(table.io())) {
+                if (validMetadataFilePaths.contains(manifest.path())) {
+                    // Already read this manifest
+                    continue;
+                }
+
+                validMetadataFilePaths.add(manifest.path());
+                if (manifest.content() == ManifestContent.DATA) {
+                    try (ManifestReader<DataFile> manifestReader = ManifestFiles.read(manifest, table.io());
+                        CloseableIterator<DataFile> dataFiles = manifestReader.iterator()) {
+                        while (dataFiles.hasNext()) {
+                            validDataFilePaths.add(dataFiles.next().path().toString());
+                        }
+                    }
+                    catch (IOException e) {
+                        throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Unable to list manifest file content from " + manifest.path(), e);
+                    }
+                }
+                else if (manifest.content() == ManifestContent.DELETES) {
+                    try (ManifestReader<DeleteFile> manifestReader = ManifestFiles.readDeleteManifest(manifest, table.io(), table.specs());
+                            CloseableIterator<DeleteFile> dataFiles = manifestReader.iterator()) {
+                        while (dataFiles.hasNext()) {
+                            validDeleteFilePaths.add(dataFiles.next().path().toString());
+                        }
+                    }
+                    catch (IOException e) {
+                        throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Unable to list manifest file content from " + manifest.path(), e);
+                    }
+                }
+                else {
+                    throw new TrinoException(ICEBERG_INVALID_METADATA, "Unknown Iceberg manifest content: " + manifest);
```

Comment:
use 

```
switch (manifest.content()) { 
case DATA -> ...
```

this will ensure the code gets appropriately updated when the library changes